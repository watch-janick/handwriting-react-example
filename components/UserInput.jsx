import React, { PureComponent } from 'react'
import PropTypes from 'prop-types'
import autobind from 'autobind-decorator'
import debounce from 'lodash.debounce'

import {
  Input,
} from 'reactstrap'

/**
 * Minimum delay between two requests
 */
const debounceRate = 400

/**
 * Component used to collect the user input and send it
 * back after each change.
 *
 * @extends PureComponent
 */
class UserInput extends PureComponent {
  static propTypes = {
    onChange: PropTypes.func.isRequired,
    defaultText: PropTypes.string,
  }

  @autobind // use autobind to make this available in textChanged
  textChanged (e) {
    const value = e.target.value

    this.onChangeDebounced(value)
  }

  /**
   * Debounced function to minimize the number of
   * requests sent to the API.
   */
  onChangeDebounced = debounce(value => {
    const { onChange } = this.props

    onChange(value)
  }, debounceRate)

  render () {
    const { defaultText } = this.props

    return (
      <Input
        type="textarea"
        name="text"
        id="user-input"
        defaultValue={defaultText}
        onChange={this.textChanged}
      />
    )
  }
}

export default UserInput