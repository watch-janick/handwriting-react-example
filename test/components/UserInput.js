import UserInput from '../../components/UserInput'

describe('UserInput', () => {
  it('should render correctly with defaultText', () => {
    const attrs = {
      defaultText: 'this is a default text',
      onChange: fn,
    }
    const component = shallow(<UserInput {...attrs} />)

    expect(component).to.matchSnapshot()
  })

  it('should call onChange when text changes', () => {
    const onChange = spy()
    const value = 'this is my new text'
    const attrs = {
      defaultText: 'this is a default text',
      onChange,
    }

    // fake the timers to disable the debounced function in UserInput
    // Otherwise expect is called before onChange has been called
    const clock = sinon.useFakeTimers()
    const component = shallow(<UserInput {...attrs} />)
    component.find('#user-input').simulate('change', { target: { value } } )

    // enough time for the debounce function to fire
    clock.tick(2000)

    expect(onChange.calledOnceWith(value)).to.equal(true)
  })
})