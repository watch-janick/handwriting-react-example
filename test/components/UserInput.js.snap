// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UserInput should render correctly with defaultText 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <UserInput
    defaultText="this is a default text"
    onChange={[Function]}
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "class",
    "props": Object {
      "defaultValue": "this is a default text",
      "id": "user-input",
      "name": "text",
      "onChange": [Function],
      "type": "textarea",
    },
    "ref": null,
    "rendered": null,
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "defaultValue": "this is a default text",
        "id": "user-input",
        "name": "text",
        "onChange": [Function],
        "type": "textarea",
      },
      "ref": null,
      "rendered": null,
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
    "attachTo": undefined,
    "hydrateIn": undefined,
  },
}
`;
